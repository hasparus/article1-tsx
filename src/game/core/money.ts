export type DragonCoins = number & { __type: "DragonCoins" };

export function DragonCoins(coins: number) {
  return coins as DragonCoins;
}

export type Gold = number & { __type?: "Gold" };

export function Gold(gold: number) {
  return gold as Gold;
}
