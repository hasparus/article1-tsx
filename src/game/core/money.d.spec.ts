import { DragonCoins, Gold } from "./money";

// Implicit conversion from numbers to Gold should work.
// @dts-jest:pass:snap
const _g1: Gold = 1000;

type NotGold = number & { __type?: "NotGold" };
// @dts-jest:fail:snap
const _g2: NotGold = Gold(1000);

// Implicit conversion from numbers to DragonCoins shouldn't work.
// @dts-jest:fail:snap
const _d1: DragonCoins = 10 ** 12;

// @dts-jest:pass:snap
const _d2: DragonCoins = DragonCoins(100);

// We can still treat DragonCoins and Gold as numbers
// @dts-jest:pass:snap
const _g3 = Gold(10000).toLocaleString();

// @dts-jest:pass:snap
const _d3 = DragonCoins(100).toFixed(2);
