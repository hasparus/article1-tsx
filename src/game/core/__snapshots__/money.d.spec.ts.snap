// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`const _d1: DragonCoins = 10 ** 12 (type) should match snapshot: const _d1: DragonCoins = 10 ** 12 1`] = `
"Type 'number' is not assignable to type 'DragonCoins'.
  Type 'number' is not assignable to type '{ __type: \\"DragonCoins\\"; }'."
`;

exports[`const _d2: DragonCoins = DragonCoins(100) (type) should match snapshot: const _d2: DragonCoins = DragonCoins(100) 1`] = `"any"`;

exports[`const _d3 = DragonCoins(100).toFixed(2) (type) should match snapshot: const _d3 = DragonCoins(100).toFixed(2) 1`] = `"any"`;

exports[`const _g1: Gold = 1000 (type) should match snapshot: const _g1: Gold = 1000 1`] = `"any"`;

exports[`const _g2: NotGold = Gold(1000) (type) should match snapshot: const _g2: NotGold = Gold(1000) 1`] = `
"Type 'Gold' is not assignable to type 'NotGold'.
  Type 'Gold' is not assignable to type '{ __type?: \\"NotGold\\"; }'.
    Types of property '__type' are incompatible.
      Type '\\"Gold\\"' is not assignable to type '\\"NotGold\\"'."
`;

exports[`const _g3 = Gold(10000).toLocaleString() (type) should match snapshot: const _g3 = Gold(10000).toLocaleString() 1`] = `"any"`;
